modules/3d/test/test_odometry.cpp

```diff
@@ -462,4 +401,246 @@ TEST(RGBD_Odometry_FastICP, prepareFrame)
+    warpedDepth = warpedDepthCvt;
+
+    Scalar badVal;
+    switch (depthType)
+    {
+    case CV_16U:
+        badVal = 0;
+        break;
+    case CV_32F:
+        badVal = std::numeric_limits<float>::quiet_NaN();
+        break;
+    case CV_64F:
+        badVal = std::numeric_limits<double>::quiet_NaN();
+        break;
+    default:
+        CV_Error(Error::StsBadArg, "Unsupported depth data type");
+    }
+
+    srcDepth.setTo(badVal, ~epsSrc);
+    warpedDepth.setTo(badVal, ~epsWarped);
+
+    if (checkMask)
+    {
+        srcMask = epsSrc; warpedMask = epsWarped;
+    }
+    else
+    {
+        srcMask = Mat(); warpedMask = Mat();
+    }
+
+    if (needRgb)
+    {
+        srcRgb = imread(srcRgbFilename, rgbType == CV_8UC1 ? IMREAD_GRAYSCALE : IMREAD_COLOR);
+        ASSERT_FALSE(srcRgb.empty()) << "Image " << srcRgbFilename.c_str() << "can not be read" << std::endl;
+
+        if (identityTransform)
+        {
+            srcRgb.copyTo(warpedRgb, epsSrc);
+        }
+        else
+        {
+            warpedRgb = imread(warpedRgbFilename, rgbType == CV_8UC1 ? IMREAD_GRAYSCALE : IMREAD_COLOR);
+            ASSERT_FALSE (warpedRgb.empty()) << "Image " << warpedRgbFilename.c_str() << "can not be read" << std::endl;
+        }
+
+        if (rgbType == CV_8UC4)
+        {
+            Mat newSrcRgb, newWarpedRgb;
+            cvtColor(srcRgb, newSrcRgb, COLOR_RGB2RGBA);
+            srcRgb = newSrcRgb;
+            // let's keep alpha channel
+            std::vector<Mat> warpedRgbChannels;
+            split(warpedRgb, warpedRgbChannels);
+            warpedRgbChannels.push_back(epsWarped);
+            merge(warpedRgbChannels, newWarpedRgb);
+            warpedRgb = newWarpedRgb;
+        }
+
+        ASSERT_TRUE(srcRgb.type() == rgbType);
+        ASSERT_TRUE(warpedRgb.type() == rgbType);
+    }
+    else
+    {
+        srcRgb = Mat(); warpedRgb = Mat();
+    }
+
+    // test data used to generate warped depth and rgb
+    // the script used to generate is in opencv_extra repo
+    // at testdata/cv/rgbd/warped_depth_generator/warp_test.py
+    double fx = 525.0, fy = 525.0,
+           cx = 319.5, cy = 239.5;
+    Matx33d K(fx,  0, cx,
+               0, fy, cy,
+               0,  0,  1);
+    cv::Affine3d rt;
+    cv::Vec3d tr(-0.04, 0.05, 0.6);
+    rt = identityTransform ? cv::Affine3d() : cv::Affine3d(cv::Vec3d(0.1, 0.2, 0.3), tr * transScaleCoeff);
+
+    warpFrame(srcDepth, srcRgb, srcMask, rt.matrix, K, dstDepth, dstRgb, dstMask);
+}
+
+
+TEST(RGBD_Odometry_WarpFrame, inputTypes)
+{
+    // [depthType, rgbType]
+    std::array<int, 5*2> types =
+    { CV_16U, CV_8UC3,
+      CV_32F, CV_8UC3,
+      CV_64F, CV_8UC3,
+      CV_32F, CV_8UC1,
+      CV_32F, CV_8UC4 };
+    const double shortl2diff = 233.983;
+    const double shortlidiff = 1;
+    const double floatl2diff = 0.038209;
+    const double floatlidiff = 0.00020004;
+    for (int i = 0; i < 5; i++)
```

Comment:
I propose to do parameterized test with GTest rather than internal loop. In case of test failure it's more obvious. It very simple to extend.